[REF] Replicate base class, but inherit from PyTorch linear operator

f-dangel opened the pull request:
This PR introduces a class `CurvatureLinearOperator`, which is almost identical to the current base class `_LinearOperator`, but inherits from `PyTorchLinearOperator`.

The next steps will be to (i) migrate each linear operator to inherit from `CurvatureLinearOperator`, rather than `_LinearOperator`, (ii) delete `_LinearOperator`.

f-dangel commented on curvlinops/_torch_base.py:
```diff
@@ -347,3 +352,344 @@ def f_scipy(X: numpy.ndarray) -> numpy.ndarray:
+            )
+
+        in_shape = [tuple(p.shape) for p in params] if in_shape is None else in_shape
+        out_shape = [tuple(p.shape) for p in params] if in_shape is None else in_shape
```
Typo: should be `out_shape`

f-dangel commented on curvlinops/_torch_base.py:
```diff
@@ -347,3 +352,344 @@ def f_scipy(X: numpy.ndarray) -> numpy.ndarray:
+        Yields:
+            Mini-batches ``(X, y)``.
+        """
+        data_iter = iter(self._data)
```
Remove `iter`

runame commented on curvlinops/_torch_base.py:
```diff
@@ -347,3 +352,344 @@ def f_scipy(X: numpy.ndarray) -> numpy.ndarray:
+                additional trailing dimension of size number of columns).
+
+        Returns: # noqa: D402
+           Result of matrix-multiplication in list format.
```
This seems to be inconsistent with the type hint `Tuple[Tensor]`.

f-dangel replied:
Fixed.

runame reviewed (approved):
LGTM!

Would be good to add unit tests, at least of all public methods.

f-dangel commented:
Added the following test: In this PR, `CurvatureLinearOperator` has two public functions: `to_device`, and `gradient_and_loss`. The `to_device` function will be deprecated soon, therefore I am not adding a test here (it is also called by the other linear operators in their deterministic checks, therefore tested indirectly). I added an explicit test which makes sure that the loss and gradient are correctly accumulated over a data set. This functionality was tested indirectly before by making sure the curvature matrix-vector products are correctly accumulated over a data set.
